Add verify check for lowest diff drain/play time requirements

Hiviexd opened the pull request:
The changes to the context are a prerequisite/necessity for this check and future ones that need to compare difficulties and their objects/control points among each other. (see checks blocked via a `🛑 ¹` in https://github.com/ppy/osu/issues/12091#issuecomment-878760791 for starters)

This check ports the following [MV check](https://github.com/Naxesss/MapsetVerifier/blob/main/src/Checks/AllModes/Spread/CheckLowestDiff.cs) and enhances it via adjusting the thresholds to match the actual requirements of each ruleset.

Hiviexd committed:
include all of a beatmapset's diffs in the verifier context

Hiviexd committed:
Add verify check for lowest diff drain time requirement

Hiviexd committed:
add test coverage

smoogipoo commented on osu.Game/Rulesets/Edit/BeatmapVerifierContext.cs:
```diff
@@ -26,11 +28,44 @@ public class BeatmapVerifierContext
             WorkingBeatmap = workingBeatmap;
             InterpretedDifficulty = difficultyRating;
+
+            beatmapsetDifficulties = new Lazy<IReadOnlyList<IBeatmap>>(() =>
```
Is it necessary for this to be a lazy? It kinda sketches me out because doing it this way looks as if to mean "we shouldn't initialise this immediately because we could be in the wrong state during the ctor".

I wouldn't expect retrieving a whole bunch of `WorkingBeatmap`s at once to be bad enough to warrant this otherwise?

bdach replied:
I'm just gonna piggyback on this because to me this is a larger conversation about how these "set-level checks" should work.

I'm not sure about this approach which is "every difficulty is going to show the same checks even if they're not related to the difficulty". I haven't really used mapsetverifier much but from screenshots that I can see it appears to separate "set-level checks" from "difficulty-level checks", and we probably want to keep that level of separation so that users can work with it better.

The big question is where do you put this stuff, then? I can think of a few places:

- this, again, uses every difficulty to show the same checks over and over
- you could show this stuff as part of the submission wizard, and have the check result look exactly like mapset verifier's (e.g. one tab for general issues, and one tab per every other difficulty)
- there's talk of bringing back some "editor select" screen, and maybe this could go live there

All of this to say is that mechanics of this aside, I'm not even wholly convinced that this implementation *approach* is even what we want to use in the end here.

smoogipoo reviewed (commented):
.

Hiviexd committed:
unlazy BeatmapsetDifficulties

bdach commented on osu.Game/Rulesets/Edit/BeatmapVerifierContext.cs:
```diff
@@ -26,11 +28,43 @@ public class BeatmapVerifierContext
+        /// </summary>
+        public readonly IReadOnlyList<IBeatmap> BeatmapsetDifficulties;
+
+        public BeatmapVerifierContext(IBeatmap beatmap, IWorkingBeatmap workingBeatmap, DifficultyRating difficultyRating = DifficultyRating.ExpertPlus, Func<BeatmapInfo, IWorkingBeatmap?>? beatmapResolver = null)
```
I don't particularly know where to begin with this but I have a negative reaction to the shape of this.

Previously this was a somewhat simple ctor. You gave it data and it stored it. It had no logic. Now, however, it's doing logic. It's also significantly more complicated because of how it's trying to jam the concept of having a "current beatmap" with "being able to access the other beatmaps" and not being particularly clear about what's what.

Like if I am to entertain this direction, the shape of this should IMO be closer to

```csharp
public BeatmapVerifierContext
{
    public IBeatmap CurrentBeatmap { get; }
    public IWorkingBeatmap CurrentWorkingBeatmap { get; }
    public DifficultyRating DifficultyRating { get; }
    public IReadOnlyCollection<IBeatmap> OtherDifficulties { get; } = [];
    public IEnumerable<IBeatmap> AllDifficulties => [CurrentBeatmap, ..OtherDifficulties];

    public BeatmapVerifierContext(...)
    {
        // dumb ctor with no logic that only stores to the fields, for tests and such
    }

    public static BeatmapVerifierContext Create(IBeatmap currentBeatmap, IWorkingBeatmap workingBeatmap, DifficultyRating difficultyRating, BeatmapManager beatmapManager)
    {
         // put all of the logic this ctor currently has here
    }
}
```

And yes I would do these renames in separate PRs because you'd probably end up touching 50 files.

bdach commented on osu.Game/Screens/Edit/Verify/IssueList.cs:
```diff
@@ -43,7 +46,7 @@ private void load(OverlayColourProvider colours)
             rulesetVerifier = beatmap.BeatmapInfo.Ruleset.CreateInstance().CreateBeatmapVerifier();
 
-            context = new BeatmapVerifierContext(beatmap, workingBeatmap.Value, verify.InterpretedDifficulty.Value);
+            context = new BeatmapVerifierContext(beatmap, workingBeatmap.Value, verify.InterpretedDifficulty.Value, beatmapInfo => beatmapManager.GetWorkingBeatmap(beatmapInfo));
```
I really have to wonder whether this isn't stopping a step short by stopping at `.GetWorkingBeatmap()`. Something tells me that sooner rather than later you're going to want to call `.GetPlayableBeatmap()` on the result of this as well, if you want to get beatmaps which are in a *usable and inspectable* state going forward.

bdach commented on osu.Game/Rulesets/Edit/Checks/CheckLowestDiffDrainTime.cs:
```diff
@@ -0,0 +1,88 @@
+            if (difficulties.Count == 0)
+                yield break;
+
+            var lowestDifficulty = difficulties.OrderBy(b => b.BeatmapInfo.StarRating).First();
```
This is subclassed and implemented by individual rulesets, so every ruleset presumably has its own separate spread rules, but nowhere in this logic do I see any actual filtering of the beatmaps being checked to the relevant ruleset? I feel like there is potential for this to be buggy and be holding difficulties to requirements they're not supposed to be held to (e.g. a catch beatmap will get a taiko drain time warning reported on it etc)

Hiviexd committed:
address review

- add TODO for refactoring verifier context ctor
- call `GetPlayableBeatmap()` in verifier context ctor
- filter diffs with relevant ruleset in check logic
- fix tests

Hiviexd committed:
fix ci

Hiviexd committed:
reduce whitespace

Hiviexd committed:
use `TimeSpan` for defining thresholds

Hiviexd commented on osu.Game/Rulesets/Edit/Checks/CheckLowestDiffDrainTime.cs:
```diff
@@ -0,0 +1,90 @@
+        public class IssueTemplateTooShort : IssueTemplate
+        {
+            public IssueTemplateTooShort(ICheck check)
+                : base(check, IssueType.Problem, "With a lowest difficulty {0}, the {1} time of {2} must be at least {3}, currently {4}.")
```
For reference, the current wording is [taken directly from Mapset Verifier](https://github.com/Naxesss/MapsetVerifier/blob/main/src/Checks/AllModes/Spread/CheckLowestDiff.cs#L53), so it gets bonus points in terms of user familiarity.

However, I do agree that mentioning the current diffname makes it read a bit weirdly. I've opted to go with the following wording to make it more coherent while keeping the actual intention of the check:

> With the lowest difficulty being "Hard", the play time of this difficulty must be at least 3:30, currently 2:00.

re: your proposed wording, it doesn't fully make sense here since the check isn't only for checking the lowest difficulty — it references the lowest difficulty to check for drain times issues of any given difficulty.

Hiviexd committed:
adjust wording

Hiviexd committed:
oops

bdach reviewed (approved):
haven't checked the logic very closely, just going to hope that it matches the rc expectations and that the tests are good, if it doesn't it can get fixed in post
